File: app/src/main/java/com/example/solar_alarm/BroadcastReceiver/AlarmBroadcastReceiver.java
package com.example.solar_alarm.BroadcastReceiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.widget.Toast;

import com.example.solar_alarm.Service.AlarmService;
import com.example.solar_alarm.Service.RescheduleAlarmService;

import java.util.Calendar;

public class AlarmBroadcastReceiver extends BroadcastReceiver {
    public static final String MONDAY = "MONDAY";
    public static final String TUESDAY = "TUESDAY";
    public static final String WEDNESDAY = "WEDNESDAY";
    public static final String THURSDAY = "THURSDAY";
    public static final String FRIDAY = "FRIDAY";
    public static final String SATURDAY = "SATURDAY";
    public static final String SUNDAY = "SUNDAY";
    public static final String RECURRING = "RECURRING";
    public static final String TITLE = "TITLE";

    @Override
    public void onReceive(Context context, Intent intent) {
        if (Intent.ACTION_BOOT_COMPLETED.equals(intent.getAction())) {
            String toastText = String.format("Alarm Reboot");
            Toast.makeText(context, toastText, Toast.LENGTH_SHORT).show();
            startRescheduleAlarmsService(context);
        }
        else {
            String toastText = String.format("Alarm Received");
            Toast.makeText(context, toastText, Toast.LENGTH_SHORT).show();
            if (!intent.getBooleanExtra(RECURRING, false)) {
                startAlarmService(context, intent);
            } {
                if (alarmIsToday(intent)) {
                    startAlarmService(context, intent);
                }
            }
        }
    }

    private boolean alarmIsToday(Intent intent)
    {
        boolean alarmIsToday = false;
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        int today = calendar.get(Calendar.DAY_OF_WEEK);

        switch(today)
        {
            case Calendar.MONDAY:
                if (intent.getBooleanExtra(MONDAY, false))
                    alarmIsToday = true;
            case Calendar.TUESDAY:
                if (intent.getBooleanExtra(TUESDAY, false))
                    alarmIsToday = true;
            case Calendar.WEDNESDAY:
                if (intent.getBooleanExtra(WEDNESDAY, false))
                    alarmIsToday = true;
            case Calendar.THURSDAY:
                if (intent.getBooleanExtra(THURSDAY, false))
                    alarmIsToday = true;
            case Calendar.FRIDAY:
                if (intent.getBooleanExtra(FRIDAY, false))
                    alarmIsToday = true;
            case Calendar.SATURDAY:
                if (intent.getBooleanExtra(SATURDAY, false))
                    alarmIsToday = true;
            case Calendar.SUNDAY:
                if (intent.getBooleanExtra(SUNDAY, false))
                    alarmIsToday = true;
        }

        return alarmIsToday;
    }

    private void startAlarmService(Context context, Intent intent) {
        Intent intentService = new Intent(context, AlarmService.class);
        intentService.putExtra(TITLE, intent.getStringExtra(TITLE));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(intentService);
        } else {
            context.startService(intentService);
        }
    }

    private void startRescheduleAlarmsService(Context context) {
        Intent intentService = new Intent(context, RescheduleAlarmService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(intentService);
        } else {
            context.startService(intentService);
        }
    }
}

File: app/src/main/java/com/example/solar_alarm/BroadcastReceiver/AlarmBroadcastReceiver.kt
package com.example.solar_alarm.BroadcastReceiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import android.util.Log
import com.example.solar_alarm.Data.Tables.SolarAlarm
import com.example.solar_alarm.Service.AlarmService
import java.util.Calendar

class AlarmBroadcastReceiver : BroadcastReceiver() {
    private lateinit var solarAlarm: SolarAlarm

    override fun onReceive(context: Context, intent: Intent) {
        try {
            // 1. Initialize data
            val dd = getSolarAlarmFromIntent(intent)

            if (dd != null)
            {
                solarAlarm = dd
                Log.d("ALARM_DEBUG", "Received alarm: ${solarAlarm.Name}")

                // 2. Logic check: If it's recurring, check if today is an active day
                if (solarAlarm.Recurring && !alarmIsToday()) {
                    Log.d("ALARM_DEBUG", "Alarm skipped: Not scheduled for today.")
                    return
                }

                // 3. Start the AlarmService to handle notification, music and full screen intent
                val serviceIntent = Intent(context, AlarmService::class.java).apply {
                    putExtra("SolarAlarm", solarAlarm)
                }

                context.startForegroundService(serviceIntent)
            }
        } catch (e: Exception) {
            Log.e("ALARM_DEBUG", "Error in onReceive: ${e.message}")
        }
    }

    private fun alarmIsToday(): Boolean {
        val calendar = Calendar.getInstance()
        val today = calendar[Calendar.DAY_OF_WEEK]

        return when (today) {
            Calendar.MONDAY    -> solarAlarm.Monday
            Calendar.TUESDAY   -> solarAlarm.Tuesday
            Calendar.WEDNESDAY -> solarAlarm.Wednesday
            Calendar.THURSDAY  -> solarAlarm.Thursday
            Calendar.FRIDAY    -> solarAlarm.Friday
            Calendar.SATURDAY  -> solarAlarm.Saturday
            Calendar.SUNDAY    -> solarAlarm.Sunday
            else               -> false
        }
    }

    companion object {
        fun getSolarAlarmFromIntent(intent: Intent): SolarAlarm?
        {
            val solarAlarm = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
            {
                intent.getParcelableExtra("SolarAlarm", SolarAlarm::class.java)
            }
            else
            {
                @Suppress("DEPRECATION")
                intent.getParcelableExtra<SolarAlarm>("SolarAlarm")
            }
            return solarAlarm
        }
    }
}
